### ISSUE TITLE
seqan::ParseError when genotyping

### ISSUE THREAD

**Issue by songtaogui** [bkehr/popins#4]
Hi,
I got error similar to #2 when running `genotype`:

```
[PopIns 2018-12-18 21:47:15] Genotyping VCF records in 'insertions.vcf' for Q417.
terminate called after throwing an instance of 'seqan::ParseError'
  what():  Not enough values in a line.
test_run.sh: line 5:  9611 Aborted                 (core dumped) popins genotype -p . -c test_surpercontigs.fa -r ref.fa Q417
```
I was using `seqan v2.2.0` and `samtools v1.6`. The cmds were:

```shell
# assembly and merge were done before
popins contigmap -p . -c test_surpercontigs.fa -r ref.fa -t 10 -m 2G Q417
popins place-refalign -p . -c test_surpercontigs.fa -r ref.fa --readLength 150
popins place-splitalign -p . -c test_surpercontigs.fa -r ref.fa --readLength 150 Q417
popins place-finish -p . -r ref.fa
popins genotype -p . -c test_surpercontigs.fa -r ref.fa Q417
# nothing went wrong until the genotype step
```
Here is the `insertions.vcf` outside the sample dir: [insertions.vcf.gz](https://github.com/bkehr/popins/files/2690853/insertions.vcf.gz)
Below are file sizes of the sample dir `Q417`, the insertion.vcf file in the sample dir only contains header records.  

size|file
-- | --
258968468 | contigs.fa
8191 | insertions.vcf
6713468 | locations_placed.txt
168112832 | locations.txt
7232301 | locations_unplaced.txt
984980416 | non_ref.bam
9299154139 | non_ref_new.bam
70617480 | non_ref_new.bam.bai
2322634640 | paired.1.fastq
2317881228 | paired.2.fastq
97 | POPINS_SAMPLE_INFO
3125470660 | single.fastq

Any suggestions?

Any help would be appreciated.

Songtao Gui

**Comment by bjarnivh** [comment]
Can you gunzip insertions.vcf.gz and try running with -i insertions.vcf (pointing to the insertions file outside the sample dir).  I suspect popIns is trying to run the insertions.vcf file inside the sample dir.

**Comment by songtaogui** [comment]
Hi,
Thanks for the quick reply.
The `insertions.vcf` was gziped for uploading to github.
I rerun the genotype cmd with absolute path of the outside insertions.vcf :
`popins genotype -p . -c test_surpercontigs.fa -r ref.fa -i /home/gst/work/popins_test/insertions.vcf Q417`
and still got the error:

```
[PopIns 2018-12-18 22:53:25] Opening input files.
[PopIns 2018-12-18 22:53:30] Genotyping VCF records in '/home/gst/work/popins_test/insertions.vcf' for Q417.
terminate called after throwing an instance of 'seqan::ParseError'
  what():  Not enough values in a line.
Aborted (core dumped)
```

**Comment by bjarnivh** [comment]
I suspect this is due to a missing column in the vcf file. 
Can you try running this command on the vcf file:
zcat insertions.vcf.gz | awk 'BEGIN{OFS="\t"}{if($1 == "#CHROM") $9 = "FORMAT"; if($1 !~ /#/ ) $9 ="."; print $0;}' > insertions_edit.vcf
and then run the command with -i insertions_edit.vcf

**Comment by songtaogui** [comment]
Hi, @bjarnivh 

Thank you for the kind help. 
Adding `FORMAT` column solves the problem. However, the genotyping step added a new `FORMAT` column before the raw manually added column, which resulted in an extra column between `FORMAT` and `Sample(Q417)` :

#CHROM | POS | ID | REF | ALT | QUAL | FILTER | INFO | FORMAT | Q417 |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 32341 | 1:32341:FP | G | ]NODE_752654_length_104_cov_2.000000f:77]G | . | . | AR=2;AS=0.4;GS=1 | GT:PL | . | 0/1:69,0,255
1 | 42296 | 1:42296:FP | A | ]NODE_395641_length_100_cov_2.000000f:43]A | . | . | AR=2;AS=1;GS=1 | GT:PL | . | 0/1:51,0,255
1 | 55663 | 1:55663:FP | A | A[NODE_102139_length_1052_cov_15.453422f:209[ | . | . | AR=12;AS=1;GS=1 | GT:PL | . | 1/1:255,120,0
1 | 57289 | 1:57289:FP | A | ]NODE_731905_length_803_cov_6.252802f:775]A | . | . | AR=2;AS=1;GS=1 | GT:PL | . | 0/1:255,0,255
1 | 80865 | 1:80865:FP | A | ]NODE_363422_length_1162_cov_10.749570f:1135]A | . | . | AR=2;AS=0.5;GS=1 | GT:PL | . | 0/1:255,0,255
1 | 94175 | 1:94175:FP | T | T[NODE_488579_length_104_cov_2.000000r:105[ | . | . | AR=2;AS=1;GS=1 | GT:PL | . | 0/1:22,0,255
1 | 128279 | 1:128279:FP | C | ]NODE_571885_length_392_cov_6.846939f:336]C | . | . | AR=2;AS=0.666667;GS=1 | GT:PL | . | 1/1:255,26,0
1 | 171774 | 1:171774:FP | A | A[NODE_766194_length_257_cov_4.015564f:190[ | . | . | AR=2;AS=0.666667;GS=7 | GT:PL | . | 0/1:255,0,255
1 | 171428 | 1:171428:FP | C | ]NODE_422651_length_299_cov_4.695652r:84]C | . | . | NOANCHOR | GT:PL | . | 0/0:0,44,255

It seems that manually added FORMAT column is not used when genotyping although it eliminates the error.


And I have other questions for the result vcf format. 

I would like to extract the excat ALT sequences based on the result vcf. However I can't find the mate infomation in the result vcf, so I can't decide the other bondary of the given contigID in ALT field. Did I miss something? 

Thank you for your patience.

**Comment by bjarnivh** [comment]
For now you can remove the extra FORMAT column, we will get this problem fixed in the next release.   
There is one VCF record for each end of the insertion.  If you only find a single VCF record for a contig we can only tell you that where one of the ends gets inserted.  To interpret the format:

1 | 171774 | 1:171774:FP | A | A[NODE_766194_length_257_cov_4.015564f:190[
An 'A' on position 171774 on chr 1 is replaced with the sequence starting at position 190 in contig NODE_766194_length_257_cov_4.015564.

The '[' represents and insert to the right of the position and a ']' an insertion from the left of the position.

The 'f' at the end of the contig name in the variant record represents that you read the contig in the forward direction.  The 'r' represents that you read in the reverse complement direction.

**Comment by songtaogui** [comment]
Some of the vcf results contains mate BND :

#CHROM | POS | ID | REF | ALT | QUAL | FILTER | INFO |
-- | -- | -- | -- | -- | -- | -- | -- |
9 | 41801027 | 9:41801027:FP | A | A[NODE_317_length_325_cov_8.083076f:204[ | . | . | AR=9;AS=0.9;GS=1
9 | 41801016 | 9:41801016:FP | T | ]NODE_317_length_325_cov_8.083076f:371]T | . | . | NOANCHOR;SR=1

but when one contig has more than 2 BND records, the mate relationships become fuzzy:

#CHROM | POS | ID | REF | ALT | QUAL | FILTER | INFO |
-- | -- | -- | -- | -- | -- | -- | -- |
8 | 84552727 | 8:84552727:FP | G | G[NODE_295_length_313_cov_9.853035f:244[ | . | . | AR=2;AS=0.333333;GS=1
3 | 170876090 | 3:170876090:FP | A | A[NODE_295_length_313_cov_9.853035r:1[ | . | . | NOANCHOR;SR=5
3 | 170876103 | 3:170876103:FP | G | ]NODE_295_length_313_cov_9.853035r]G | . | . | AR=2;AS=0.333333
8 | 84552714 | 8:84552714:FP | C | ]NODE_295_length_313_cov_9.853035f:357]C | . | . | NOANCHOR;SR=1

Is it resonable to consider two nearest BND records of one contig as mated records?

Would it be possible to add "MATEID=mate_id" to INFO field of the result vcf in the future version?

Thank you.

**Comment by bjarnivh** [comment]
Yes, the relationship may become fuzzy and PopIns does not try to resolve this fuzzy relationship.  PopIns believes that there are more than one possible insertion sites, possibly because of repeats in the reference and possibly different parts of the contigs are inserted into different places.  

Using the two closest locations is not unreasonable.  It would also be reasonable to interpret the above example as two insert locations.

You can refer to our paper, https://www.nature.com/articles/ng.3801,  as to how we analyzed variants with multiple insert locations.

**Comment by bkehr** [comment]
Dear Songtao Gui,

I am trying to reproduce your error. Could you send me the output of the command 

   `popins genotype -h`

Thanks,
Birte

**Comment by songtaogui** [comment]
Dear bkehr,

Below is the help message of `popins genotype -h`:
 
```
popins genotype - Genotyping a sample for insertions.
=====================================================

SYNOPSIS
    popins genotype [OPTIONS] SAMPLE_ID

DESCRIPTION
    Computes genotype likelihoods for a sample for all insertions given in the input VCF file by aligning all reads,
    which are mapped to the reference genome around the insertion breakpoint or to the contig, to the reference and to
    the alternative insertion sequence. VCF records with the genotype likelihoods in GT:PL format for the individual
    are written to a file insertions.vcf in the sample directory.

REQUIRED ARGUMENTS
    SAMPLE_ID STRING

OPTIONS
    -h, --help
          Display the help message.
    --version-check BOOL
          Turn this option off to disable version update notifications of the application. One of 1, ON, TRUE, T, YES,
          0, OFF, FALSE, F, and NO. Default: 1.
    -H, --fullHelp
          Display the help message with the full list of options.
    --version
          Display version information.

  Input/output options:
    -p, --prefix STRING
          Path to the sample directories. Default: '.'.
    -i, --insertions INPUT_FILE
          Name of VCF input file. Valid filetype is: .vcf. Default: insertions.vcf.
    -c, --contigs INPUT_FILE
          Name of supercontigs file. Valid filetypes are: .fna, .fasta, and .fa. Default: supercontigs.fa.
    -r, --reference INPUT_FILE
          Name of reference genome file. Valid filetypes are: .fna, .fasta, and .fa. Default: genome.fa.

  Algorithm options:
    -m, --model STRING
          Model used for genotyping. One of DUP and RANDOM. Default: RANDOM.
    -w, --window INTEGER
          Region window size. Default: 50.

VERSION
    Last update: Thu May 24 10:43:36 2018 +0200
    popins genotype version: damp_v1-151-g4010f61
    SeqAn version: 2.4.0
```
The popins was compiled with seqan v2.4.0 for the first time. Then I noticed that seqan v2.2.0 should be used, so I removed the `popins` executable file and re-compile with seqan v2.2.0. However, I noticed that the seqan version showed in the popins help messages was still v2.4.0, now I'm not sure if the bug was caused by the wrong seqan version.

Thank you for your help!

**Comment by bkehr** [comment]
You are correct, the bug is caused by the wrong version of SeqAn. Your popins binary is compiled with SeqAn 2.4.0 as indicated in the last line of the help message.

Where did you download the two versions of SeqAn to? And did you change the Makefile?

**Comment by songtaogui** [comment]
I did not change the Makefile. I removed the popins binary, and replaced the `seqan` dir (copied from Seqan-v2.4.0/include/seqan) with seqan-v2.2.0/include/seqan, and redo `make`.

I will remove the total popins dir and install again to see if the bug was gone. 

Sorry for misleading.

**Comment by bkehr** [comment]
`make clean`
`make`

may do the job in this case.

**Comment by songtaogui** [comment]
Sorry for the late reply, re-complie with seqan-v2.2.0 does solve the problem.
Thank you.

And I have another question, the `locations.txt` file for each sample contains the candidate anchor regions on ref and the number of supported read-pairs. However, I would like to also know the corresponding regions on the query-sequence and the detail read-id of the supported read-pairs. Do you have any suggestions?

Thank you again,
Best wishes,
Songtao Gui

**Comment by bkehr** [comment]
The corresponding region is either the left or the right end of the contig depending on the orientations indicated in the locations file.

The file does not store the exact contig positions and neither the read IDs.  If you need to get this information, you currently need to go back to the file that PopIns uses to extract the locations: `sampledir/non_ref_new.bam`. The reason for not reporting more details in the locations file is space: When analyzing many samples jointly, there may be several thousand read pairs supporting a single achoring location so reporting all read IDs would considerably inflate the size of the final locations file. Could you give me a hint what you need this information for? If it is an application relevant to many people we may consider implementing an option to report more details.

**Comment by songtaogui** [comment]
I have aligned a draft assembly ( assemblied using ~20X WGS PE reads ) to a ref-genome and identified some "non-ref insertion-breakpoints" by parsing the alignments, and I want to get the breakpoints which are also supported by poorly-mapped-read-pairs and genotype them use `popins`. In order to filter the breakpoints, I would like to know which part of the corresponding region (left or right) of the query-seq were anchored to the ref-region in popins. 

I have tried to retrieve the reads by extract the anchored-regions from `non_ref_new.bam`. However, the number of retrieved read-pairs were sometimes more than the number of supported read-pairs showed in `locations.txt`, I thought the "extra read-pairs" were those failed the filter criterion during `popins contigmap`, and it is a real challenge for me to filter these reads out.

Is it reasonable that I use `insertions.vcf` generated elsewhere as `popins genotype` inputs ? And since `popins genotype` gives an `AR` INFO to each records, is it reasonable that I simply using all the breakpoints (without pre-filtering) as `genotype` input and filter the outputs by the `AR` INFO ?

Any advice and suggestion will be greatly appreciated.
